src/core/generate.ts

```diff
@@ -67,10 +68,23 @@ export function generate({
   const sourceFile = resolveModules(sourceText);
 
   // Extract the nodes (interface declarations & type aliases)
-  const nodes: Array<
-    ts.InterfaceDeclaration | ts.TypeAliasDeclaration | ts.EnumDeclaration
-  > = [];
+  const nodes: Array<TypeNode> = [];
 
+  // declare a map to store the interface name and its corresponding zod schema
+  const typeNameMapping = new Map<string, TypeNode>();
+
+  const typesNeedToBeExtracted = new Set<string>();
+
+  const typeNameMapBuilder = (node: ts.Node) => {
+    if (
+      ts.isInterfaceDeclaration(node) ||
+      ts.isTypeAliasDeclaration(node) ||
+      ts.isEnumDeclaration(node)
```

Comment:
Let’s extract this into a `isTypeNode` function, this will avoid the logic duplication and the casting bellow.

```ts
function isTypeNode(node: ts.Node): node is TypeNode {
  return (
    ts.isInterfaceDeclaration(node) ||
    ts.isTypeAliasDeclaration(node) ||
    ts.isEnumDeclaration(node)
  );
}
```

Reply:
This will cause type error. The type narrowing will not working.

Reply:
Done!